Mathlib/Tactic/Linter/Style.lean

```diff
@@ -467,4 +469,47 @@ initialize addLinter doubleUnderscore
 
 end Style.nameCheck
 
+/-! # The "openClassical" linter -/
+
+/-- The "openClassical" linter emits a warning on `open Classical` statements which are not
+scoped to a single declaration. A non-scoped `open Classical` can hide that some theorem statements
+would be better stated with explicit decidability statements.
+-/
+register_option linter.style.openClassical : Bool := {
+  defValue := false
+  descr := "enable the openClassical linter"
+}
+
+namespace Style.openClassical
+
+/-- If `stx` is syntax describing an `open` command, `extractOpenNames stx`
+returns an array of the syntax corresponding to the opened names,
+omitting any renamed or hidden items. -/
+def extractOpenNames : Syntax → Array Syntax
+  | `(command|open $arg hiding $_*)    => #[arg]
+  | `(command|open $arg renaming $_,*) => #[arg]
+  | `(command|open $arg ($_*))         => #[arg]  -- `openOnly`, hence the parens in `($_*)`
+  | `(command|open $args*)             => args
+  | `(command|open scoped $args*)      => args
+  | _ => #[]
+
+@[inherit_doc Mathlib.Linter.linter.style.openClassical]
+def openClassicalLinter : Linter where run stx := do
+    unless Linter.getLinterValue linter.style.openClassical (← getOptions) do
+      return
+    if (← get).messages.hasErrors then
+     return
+    -- If `stx` describes an `open` command, extract the list of opened namespaces.
+    for stxN in (extractOpenNames stx).filter (·.getId == `Classical) do
+      Linter.logLint linter.style.openClassical stxN "\
+      please avoid 'open (scoped) Classical' statements: this can hide theorem statements \
+      which would be better stated with explicit decidability statements.\n\
+      Instead, use `open Classical in` for definitions or instances, the `classical` tactic \
+      for proofs.\nFor theorem statements, \
+      either add missing decidability assumptions or use `open Classical in`."
```

Comment:
```suggestion
    let scope ← getScope
    if scope.openDecls.any (match · with
      | .simple `Classical _ => true
      | .explicit `Classical _ => true
      | _ => false ) then
      Linter.logLint linter.style.openClassical stx "please avoid […]"
```

Just a suggestion: Would something like this be more stable? inspired by `#where`.

Note that this snippet currently lints everything after the first bare `open Classical`. Maybe that could be improve this snippet?

Reply:
Thanks for asking - this is certainly worth a thought. I don't think this would be better behaviour, though:
- this fires on `open Classical in` - which we do want to allow
- if I `open Classical` in a section, this should fire on every command afterwards, right?

I think warning just once at the `open` syntax is nicer to users - so would be inclined to leave this as-is.

Reply:
I have added a test for the second point, documenting that the linter only fires on the bad open statement itself.

Reply:
Thanks!